src/utils/sources.ts

```diff
@@ -0,0 +1,159 @@
+import {NamePath} from 'antd/lib/form/interface';
+
+import {Option} from '@models/form';
+import {SourceWithRepository} from '@models/sources';
+
+import {notificationCall} from '@molecules';
+
+const customGitSourceString = '$custom-git-';
+
+export const remapTestSources = (testSources: SourceWithRepository[]) => {
+  if (!testSources || !testSources.length) {
+    return [];
+  }
+
+  return testSources.map(source => {
+    const {name} = source;
+
+    const optionValue = `${customGitSourceString}${name}`;
+    const optionName = `Git source: ${name}`;
+
+    return {...source, value: optionValue, label: optionName};
+  });
+};
+
+export const testSourceBaseOptions: Option[] = [
+  {value: 'git', label: 'Git'},
+  {value: 'file-uri', label: 'File'},
+  {value: 'string', label: 'String'},
+];
+
+export const getTestSourceSpecificFields = (values: any, isCustomGit?: boolean) => {
+  const {testSource, username, token, string, file, ...rest} = values;
+
+  if (rest.commit) {
+    rest.branch = '';
+  }
+
+  if (rest.branch) {
+    rest.commit = '';
+  }
+
+  if (isCustomGit) {
+    return {
+      repository: {...rest, uri: '', type: ''},
+    };
+  }
+
+  if (testSource === 'string' || testSource === 'file-uri') {
+    return {data: string || file.fileContent, repository: {}};
+  }
+
+  const secrets: any = {};
+
+  if (token !== undefined && token !== null) {
+    if (token === '') {
+      secrets.tokenSecret = {};
+    } else if (!token.includes('*')) {
+      secrets.token = token;
+    }
+  }
+
+  if (username !== undefined) {
+    if (username === '') {
+      secrets.usernameSecret = {};
+    } else if (!username.includes('*')) {
```

Comment:
What if someone actually puts `*` in a username or token - shouldn't it instead show a validation error?